kore/src/Kore/Builtin/Bool.hs

```diff
@@ -165,3 +174,62 @@ builtinFunctions =
         Builtin.binaryOperator extractBoolDomainValue asPattern
     xor a b = (a && not b) || (not a && b)
     implies a b = not a || b
+
+termAndEquals
+    :: forall variable unifier
+    .  InternalVariable variable
+    => MonadUnify unifier
+    => TermSimplifier variable unifier
+    -> TermLike variable
+    -> TermLike variable
+    -> MaybeT unifier (Pattern variable)
+termAndEquals unifyChildren a b =
+    worker a b <|> worker b a
+  where
+    unifyChildren' term1 term2 =
+        Pattern.withoutTerm <$> unifyChildren term1 term2
+    worker termLike1 termLike2
+      | Just value1 <- matchBool termLike1
+      , Just value2 <- matchBool termLike2
+      = lift $ if value1 == value2
+            then return (Pattern.fromTermLike termLike1)
+            else
+                Unify.explainAndReturnBottom
+                    "different Bool domain values"
+                    termLike1
+                    termLike2
+      | Just value1 <- matchBool termLike1
+      , value1
+      , Just BoolAnd { operand1, operand2 } <- matchBoolAnd termLike2
+      , isFunctionPattern termLike2
+      = lift $ do
+        unification1 <- unifyChildren' termLike1 operand1
+        unification2 <- unifyChildren' termLike1 operand2
+        let conditions = unification1 <> unification2
+        pure (Pattern.withCondition termLike1 conditions)
+
+    worker _ _ = empty
+
+{- | Match a @BOOL.Bool@ builtin value.
+ -}
+matchBool :: TermLike variable -> Maybe Bool
+matchBool (BuiltinBool_ Domain.InternalBool { builtinBoolValue }) =
+    Just builtinBoolValue
+matchBool _ = Nothing
+
+{- | The @BOOL.and@ hooked symbol applied to @term@-type arguments.
+ -}
+data BoolAnd term =
+    BoolAnd
+        { symbol :: !Symbol
+        , operand1, operand2 :: !term
+        }
+
+{- | Match the @BOOL.and@ hooked symbol.
+ -}
+matchBoolAnd :: TermLike variable -> Maybe (BoolAnd (TermLike variable))
```

Comment:
So this kind of constructs would be used instead of pattern synonyms?
I could get used to that...

Reply:
Yes. I wanted to use a pattern synonym, but then I can't use the guard to check that the symbol is hooked to `BOOL.and`. (Actually, there is a way to "fake" guards using view patterns and tuples, but that starts to get complicated fast.)